src/main/java/MayoBot/task/TaskList.java

```diff
@@ -0,0 +1,222 @@
+     * appropriate indentation for visual consistency.
+     * <p>
+     * If the list is empty, no output is produced. The display format
+     * includes the task number, followed by the task's string representation.
+     */
+    public void printTasks(Ui ui) {
+        for (int i = 0; i < tasks.size(); i++) {
+            Task task = tasks.get(i);
+            ui.showMessage((i + 1) + ". " + task);
+        }
+    }
+
+    /**
+     * Marks the specified task as completed and updates storage.
+     * Changes the completion status of the task at the given one-based index
+     * to done and saves the entire task list to storage. Performs bounds
+     * checking to ensure the index is valid.
+     * <p>
+     * Returns a boolean indicating success or failure, allowing callers to
+     * handle invalid indices appropriately. On success, the change is
+     * immediately persisted to storage.
+     *
+     * @param index the one-based index of the task to mark as done
+     * @return true if the task was successfully marked, false if index is invalid
+     */
+    public boolean markTaskAsDone(int index) {
+        if (index <= tasks.size()) {
+            tasks.get(index - 1).markAsDone();
+            storage.saveTasks(this);
+            return true;
+        } else {
+            return false;
+        }
+
+    }
+
+    /**
+     * Marks the specified task as not completed and updates storage.
+     * Changes the completion status of the task at the given one-based index
+     * to not done and saves the entire task list to storage. Performs bounds
+     * checking to ensure the index is valid.
+     * <p>
+     * Returns a boolean indicating success or failure, allowing callers to
+     * handle invalid indices appropriately. On success, the change is
+     * immediately persisted to storage.
+     *
+     * @param index the one-based index of the task to mark as not done
+     * @return true if the task was successfully unmarked, false if index is invalid
+     */
+    public boolean markTaskAsNotDone(int index) {
+        if (index <= tasks.size()) {
+            tasks.get(index - 1).markAsNotDone();
+            storage.saveTasks(this);
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    /**
+     * Finds and returns all tasks that contain the specified search term.
+     * Searches through all task descriptions for the given keyword and returns
+     * matching tasks with their original numbering. The search is case-insensitive
+     * and matches partial strings within task descriptions.
+     * <p>
+     * Returns an empty list if no tasks match the search term. The returned
+     * indices correspond to the original task positions in the main task list.
+     *
+     * @param searchTerm the keyword to search for in task descriptions
+     * @return an ArrayList containing arrays of [originalIndex, matchingTask]
+     */
+    public ArrayList<Object[]> findTask(String searchTerm) {
```

Comment:
It may be better to use the plural form for this as the return type is an array.
```suggestion
    public ArrayList<Object[]> findTasks(String searchTerm) {
```